generate.py

```diff
@@ -0,0 +1,172 @@
+    # generate a multi track sequence
+    generated_multi_track_sequence = gen.generate_multi_track_sequence(
+        inst_list=["INST=DRUMS", "INST=38", "INST=82", "INST=51"],
+        density_list=[3, 6, 2, 1])
+    """
+
+    def __init__(
+        self,
+        model,
+        tokenizer,
+        device="cpu",
+        max_seq_length=2048,
+        temperature=0.75,
+        generate_until="TRACK_END",
+    ):
+        self.model = model
+        self.tokenizer = tokenizer
+        self.device = device
+        self.max_length = max_seq_length
+        self.temperature = temperature
+        self.generate_until = generate_until
+
+    def tokenize_input(self, input, verbose=False):
+        input_ids = self.tokenizer.encode(input, return_tensors="pt")
+        if self.device == "cuda":  # TO CHECK - not sure if it works
+            input_ids.cuda()
+
+        if verbose:
+            print(f"inputs: {self.tokenizer.decode(input_ids[0])}")
+        return input_ids
+
+    # generate from the tokenized input
+    def generate_the_next_8_bars(
+        self,
+        input_ids,
+        verbose=False,
+    ):
+        generated_ids = self.model.generate(
+            input_ids,
+            max_length=self.max_length,
+            do_sample=True,
+            temperature=self.temperature,
+            eos_token_id=self.tokenizer.encode(self.generate_until)[0],
+        )
+        if verbose:
+            print(f"output: {generated_ids}")
+        return generated_ids
+
+    # convert the generated tokens to string
+    def convert_ids_to_text(self, generated_ids, verbose=False):
+        generated_text = self.tokenizer.decode(generated_ids[0])
+        if verbose:
+            print(f"output: {generated_text}")
+        return generated_text
+
+    def generate_one_sequence(self, input="PIECE_START", inst=None, density=None):
+        if inst is not None:  # ex: inst="INST=DRUMS"
+            input = f"{input} TRACK_START {inst} "
+            if density is not None:  # ex: inst="INST=DRUMS"
+                input = f"{input} DENSITY={density}"
+        if inst is None and density is not None:
+            print("Density cannot be defined without an input instrument #TOFIX")
+
+        input_ids = self.tokenize_input(input)
+        generated_ids = self.generate_the_next_8_bars(input_ids)
+        generated_text = self.convert_ids_to_text(generated_ids)
+        return generated_text
+
+    def generate_multi_track_sequence(
+        self,
+        inst_list=["INST=DRUMS", "INST=38", "INST=82"],
+        density_list=[3, 6, 2],
+    ):
+        generate_features_dict = {
+            "inst_list": inst_list,
+            "density_list": density_list,
+            "temperature": self.temperature,
+            "max_seq_length": self.max_length,
+            "generate_until": self.generate_until,
+        }
+
+        generated_multi_track_sequence = []
+        input = "PIECE_START"
```

Comment:
it might be a better idea not to use input as variable name, it is reserved for another builtin function, my compiler complains